irohad/multi_sig_transactions/storage/storage_limit.hpp

```diff
@@ -0,0 +1,219 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_MST_STORAGE_GUARD_HPP
+#define IROHA_MST_STORAGE_GUARD_HPP
+
+#include <atomic>
+#include <memory>
+#include <type_traits>
+
+#include <boost/core/noncopyable.hpp>
+#include <boost/optional.hpp>
+#include "interfaces/iroha_internal/transaction_batch.hpp"
+#include "multi_sig_transactions/mst_types.hpp"
+
+namespace iroha {
+
+  struct LimitableStorage {
```

Comment:
It looks like with this PR we want to solve two tasks:
1. Limit some storage with an upper bound size
2. Make "defer" remove from the collection.

The first task can be solved by `StorageLimit` strategy and this strategy should be passed into a collection as the dependency.

The second task should be solved by inheritance at the collection from an interface with such semantics:

```
template<typename Item>
class DeferRemove {
Proxy deferRemove(Key); // Key is some tag for removing objects from the collection here it is doesn't matter finalized interface. Only proxy is important here.
}

template<typename Item>
class Proxy {

Item release(); // method removes item from initial collection
private:
Iter<Item> obj; // there could be a pointer, iterator or identifier in the collection.
}
```
With this approach, all synchronization stuff will hold in the collection and will be consistent with the rest of the collection class and code will look cleaner because we separate two responsibilities - there is no intersection between add functionality in the collection and add which bad case of removing. Also, there is no need util classes such as `MstToPcsPropagation` because all stuff is doing in one subscribe on PCS.